libjulia_jll 1.7

fingolfin opened the pull request:
This is work in progress; I figure there'll be problems, but we
will need this to rebuild CxxWrap for 1.7 (otherwise everything
using CxxWrap will segfault there)﻿

fingolfin commented:
@giordano @staticfloat @barche So this PR verifies what was said on Slack: that to build libjulia_jll 1.7, we either need to run Julia 1.7 on Yggdrasil, or else need some other new infrastructure that allows e.g. installing LLVM JLLs for 1.7 from Julia 1.6.

So, how do we proceed? Anything I / "we" can do to help expedite this?

We need this for a new Libjulia, and I am getting more and more reports by people who tried 1.7-beta2 and found that they couldn't use "my" packages (all boiling down to the CxxWrap issues).

fingolfin commented:
Note that I also tried to just build without any dependencies, on the theory that Julia builds this way on my machine, too, and is normally able to download most of what it needs. That didn't work, though, it failed compiling the Julia kernel because of some LLVm headers it couldn't find, see [this log](https://dev.azure.com/JuliaPackaging/Yggdrasil/_build/results?buildId=11894&view=logs&j=572c5e49-83d5-5271-390a-e6dc77f89c6b&t=25aa3cdb-e20d-599e-657c-8776cc8a525f).

But I wonder if this approach could be made to work anyway. Of course it's not nice to not declare all those JLLs as dependencies, *but* if it would allow us to get a first version of libjulia_jll 1.7 out, that would be enough to get going with CxxWrap etc.; and a "proper" libjulia_jll 1.7 could be done later.

vchuravy commented:
@fingolfin I can't edit that part on the web, can you adda another case here: https://github.com/JuliaPackaging/Yggdrasil/pull/3214/files#diff-29b8d08819b1a971a0a4f552117850f0569b1947bc3390d698455c6e3b7557b7R106 to use LLVM12?

vchuravy commented:
Yay!

fingolfin commented:
nice. Of course question remains whether it works. Also not optimal to call the prerelease / beta binary 1.7.0, but hey...

I'll clean up the whole thing a bit.

fingolfin commented:
Also will re-enable the other platforms, we'll see...

fingolfin commented:
- macOS ARM failure seems to be caused by `utf8proc_jll` not having a binary for that platform
- macOS Intel errors out with `ld: library not found for -losxunwind`
- Linux armv6l:  fails with error`src/flisp/string.c:22:10: fatal error: utf8proc.h: No such file or directory`

giordano commented:
aarch64-apple-darwin and armv6l-* are basically the same issue, they are the "experimental" platforms.  For macOS I think we use llvm unwind now?

fingolfin commented:
Trying to fix the x86 macos build now.

Regarding the "experimental" arm platforms, should I just disable them (for now), so we can move forward?

fingolfin commented:
Hmm, seems Yggdrasil doesn't want to build my latest pushes? Don't see them under https://dev.azure.com/JuliaPackaging/Yggdrasil/_build either

giordano commented:
Yes, also #3251 wasn't built on master (CC: @maleadt).  Agents appear to be online, I think there is something wrong with Azure Pipelines in general

fingolfin commented:
Wonder if this is related: https://status.dev.azure.com/_event/246338456

giordano commented:
I saw that, but I don't think we use artifacts, also, that started 12 days ago, so I feel like it's unrelated

giordano commented:
I manually triggered https://dev.azure.com/JuliaPackaging/Yggdrasil/_build/results?buildId=12023&view=results for CUDA and this is now working.  Let's see whether closing this pull request does the trick

giordano commented:
/AzurePipelines run

giordano commented:
Well, I don't know if it was thanks to me, but now everything started working.

fingolfin commented:
ERROR: LoadError: KeyError: key UUID("47c5dbc3-30ba-59ef-96a6-123e260183d9") not found


huh?

giordano commented:
@vchuravy @staticfloat can some of you please look into this?  I don't have access :see_no_evil:

fingolfin commented:
That UUID is `LLVMLibUnwind_jll` which for some reason is not found? I've disabled this again for now. But then still need a way to fix the macOS Intel build.

giordano commented:
Can you please point me directly to the log of the offending job?  :slightly_smiling_face:

benlorenz commented:
The error comes from something like this running in julia 1.6 (there is a platform argument in the code but that doesn't make a difference):
```
julia> ctx = Pkg.Types.Context(;julia_version=v"1.7.0");

julia> Pkg.add(ctx,[PackageSpec(name="LLVMLibUnwind_jll")])
   Resolving package versions...
ERROR: KeyError: key UUID("47c5dbc3-30ba-59ef-96a6-123e260183d9") not found
Stacktrace:
  [1] getindex
    @ ./dict.jl:482 [inlined]
  [2] deps_graph(ctx::Pkg.Types.Context, uuid_to_name::Dict{UUID, String}, reqs::Dict{UUID, Pkg.Types.VersionSpec}, fixed::Dict{UUID, Pkg.Resolve.Fixed})
    @ Pkg.Operations /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.6/Pkg/src/Operations.jl:474
  [3] resolve_versions!(ctx::Pkg.Types.Context, pkgs::Vector{Pkg.Types.PackageSpec})
    @ Pkg.Operations /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.6/Pkg/src/Operations.jl:404
  [4] targeted_resolve(ctx::Pkg.Types.Context, pkgs::Vector{Pkg.Types.PackageSpec}, preserve::Pkg.Types.PreserveLevel)
    @ Pkg.Operations /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.6/Pkg/src/Operations.jl:1210
  [5] tiered_resolve(ctx::Pkg.Types.Context, pkgs::Vector{Pkg.Types.PackageSpec})
    @ Pkg.Operations /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.6/Pkg/src/Operations.jl:1182
  [6] _resolve
    @ /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.6/Pkg/src/Operations.jl:1216 [inlined]
  [7] add(ctx::Pkg.Types.Context, pkgs::Vector{Pkg.Types.PackageSpec}, new_git::Vector{UUID}; preserve::Pkg.Types.PreserveLevel, platform::Platform)
    @ Pkg.Operations /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.6/Pkg/src/Operations.jl:1231
  [8] add(ctx::Pkg.Types.Context, pkgs::Vector{Pkg.Types.PackageSpec}; preserve::Pkg.Types.PreserveLevel, platform::Platform, kwargs::Base.Iterators.Pairs{Union{}, Union{}, Tuple{}, NamedTuple{(), Tuple{}}})
    @ Pkg.API /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.6/Pkg/src/API.jl:203
  [9] add(ctx::Pkg.Types.Context, pkgs::Vector{Pkg.Types.PackageSpec})
    @ Pkg.API /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.6/Pkg/src/API.jl:154
 [10] top-level scope
    @ REPL[60]:1
```
LLVMLibUnwind_jll is a stdlib from the future for julia 1.6 and some of the code seems unable to deal with that (and it has changed quite a bit between 1.6 and current Pkg.jl master).

benlorenz commented on L/libjulia/common.jl:
```diff
@@ -267,36 +288,58 @@ function build_julia(ARGS, version)
+        push!(dependencies, BuildDependency(get_addable_spec("LLVM_full_jll", v"12.0.0+0")))
+
+        # starting with Julia 1.7, we need LLVMLibUnwind_jll
+        #push!(dependencies, Dependency("LLVMLibUnwind_jll"))
```
```suggestion
        push!(dependencies, BuildDependency(get_addable_spec("LLVMLibUnwind_jll", v"11.0.1+1")))
```

fingolfin commented:
Hrm...
[ERROR: LoadError: invalid git HEAD (reference 'refs/heads/master' not found)](https://dev.azure.com/JuliaPackaging/Yggdrasil/_build/results?buildId=12050&view=logs&j=572c5e49-83d5-5271-390a-e6dc77f89c6b&t=25aa3cdb-e20d-599e-657c-8776cc8a525f&l=1972)

fingolfin commented:
Off-topic: in the macOS build, there are lots of warnings `make: xcode-select: No such file or directory` caused by the build system invoking `xcode-select -p`. I wonder if this should be either patched out, or if there should be a fake `xcode-select` binary which returns something sensible?

fingolfin commented:
The macOS x86 build now proceeds a bit further, but dies like [in this log](https://dev.azure.com/JuliaPackaging/Yggdrasil/_build/results?buildId=12050&view=logs&j=5850b9b5-9a51-573f-7d03-f1fedbd8a7fb&t=34cffb34-e22b-5097-6aba-496cb946d943&l=2936)
```
[20:03:43]  x86_64-apple-darwin14-clang++ -mmacosx-version-min=10.10 -m64 -shared  -pipe -fPIC -fno-rtti -pedantic -std=c++14 -D_FILE_OFFSET_BITS=64 -I/workspace/destdir/include -DNDEBUG  -O3 -g -D_GNU_SOURCE -I. -I/workspace/srcdir/julia-b570546b68/src -I/workspace/srcdir/julia-b570546b68/src/flisp -I/workspace/srcdir/julia-b570546b68/src/support -I/workspace/destdir/include -I/workspace/srcdir/julia-b570546b68/usr/include -DLIBRARY_EXPORTS -I/workspace/srcdir/julia-b570546b68/deps/valgrind -Wall -Wno-strict-aliasing -fno-omit-frame-pointer -fvisibility=hidden -fno-common -Wno-comment -Wpointer-arith -Wundef -DJL_BUILD_ARCH='"x86_64"' -DJL_BUILD_UNAME='"Darwin"' -I -DLLVM_SHLIB "-DJL_SYSTEM_IMAGE_PATH=\"../lib/julia/sys.dylib\"" "-DJL_LIBJULIA_SONAME=\"libjulia.1.dylib\""       "-DJL_LIBJULIA_INTERNAL_SONAME=\"libjulia-internal.1.dylib\"" ./jloptions.o ./runtime_ccall.o ./rtutils.o ./codegen.o ./llvm-ptls.o ./jltypes.o ./gf.o ./typemap.o ./smallintset.o ./ast.o ./builtins.o ./module.o ./interpreter.o ./symbol.o ./dlload.o ./sys.o ./init.o ./task.o ./array.o ./dump.o ./staticdata.o ./toplevel.o ./jl_uv.o ./datatype.o ./simplevector.o ./runtime_intrinsics.o ./precompile.o ./threading.o ./partr.o ./stackwalk.o ./gc.o ./gc-debug.o ./gc-pages.o ./gc-stacks.o ./method.o ./jlapi.o ./signal-handling.o ./safepoint.o ./timing.o ./subtype.o ./crc32c.o ./APInt-C.o ./processor.o ./ircode.o ./opaque_closure.o ./jitlayers.o ./aotcompile.o ./debuginfo.o ./disasm.o ./llvm-simdloop.o ./llvm-muladd.o ./llvm-final-gc-lowering.o ./llvm-pass-helpers.o ./llvm-late-gc-lowering.o ./llvm-lower-handlers.o ./llvm-gc-invariant-verifier.o ./llvm-propagate-addrspaces.o ./llvm-multiversioning.o ./llvm-alloc-opt.o ./cgmemmgr.o ./llvm-api.o ./llvm-remove-addrspaces.o ./llvm-remove-ni.o ./llvm-julia-licm.o ./llvm-demote-float16.o -Wl,-rpath,'@loader_path/julia/' -Wl,-rpath,'@loader_path/' -o /workspace/srcdir/julia-b570546b68/usr/lib/libjulia-internal.1.7.dylib -L/workspace/destdir/lib -Wl,-compatibility_version,1 -Wl,-current_version,1.7.0 -L/workspace/srcdir/julia-b570546b68/usr/lib -L/workspace/srcdir/julia-b570546b68/usr/lib -Xlinker -all_load ./flisp/libflisp.a -Xlinker -all_load ./support/libsupport.a -ljulia /workspace/destdir/lib/libuv.a -lutf8proc   -L/workspace/destdir/lib -lLLVM -framework CoreFoundation 
[20:03:43] dsymutil /workspace/srcdir/julia-b570546b68/usr/lib/libjulia-internal.1.7.dylib
[20:03:44] /opt/bin/x86_64-apple-darwin14-libgfortran4-cxx11-julia_version+1.7.0/dsymutil: line 5:  4128 Segmentation fault      "$@"
[20:03:44] make[1]: *** [Makefile:302: /workspace/srcdir/julia-b570546b68/usr/lib/libjulia-internal.1.7.dylib] Error 139
```

giordano commented:
[`e5f6b33` (#1987)](https://github.com/JuliaPackaging/Yggdrasil/pull/1987/commits/e5f6b33647f8f32a8eb4fe26b75518ee5ab287ea)

fingolfin commented:
@giordano ah thanks -- that patch did not apply anylong and I removed it, thinking "I'll see what happens if I do that", and then forgot about this when this PR took too long ;-). I've added a new equivalent patch, this time with a git commit message in it (borrowed from your PR) to indicate what it is about.

fingolfin commented:
Seems all is building. However, there are a bunch of warnings that may need to be resolved. Also, I am not sure those Dependencies, as they are, are "right" / "what we want" ? Perhaps @vchuravy @giordano @staticfloat would you like to scrutinize this some more.

vchuravy commented on L/libjulia/common.jl:
```diff
@@ -15,6 +17,12 @@ function libjulia_platforms(julia_version)
+    # HACK: remove platforms that are not building with Julia 1.7.0-beta2 right now
+    if julia_version ≥ v"1.7"
+        filter!(p -> !(Sys.islinux(p) && arch(p) == "armv6l"), platforms)
+        filter!(p -> !(Sys.isapple(p) && arch(p) == "aarch64"), platforms)
```
Is this still the case? :cry:

fingolfin replied:
This is because utf8proc is missing -- I think if someone updates https://github.com/JuliaBinaryWrappers/utf8proc_jll.jl/ to support these architectures, they might work (or at least progress a bit further). 

Actually, that should be mentioned in the comment above (I thought I had done so, but seems I forgot)

fingolfin replied:
macOS arm and Linux armv6l all fail with error `src/flisp/string.c:22:10: fatal error: utf8proc.h: No such file or directory`

giordano replied:
#3261

vchuravy commented on L/libjulia/common.jl:
```diff
@@ -77,10 +86,11 @@ function build_julia(ARGS, version)
     override OS=Linux
     EOM
 
+    LLVM_CXXFLAGS="-I${prefix}/include -fno-exceptions -fno-rtti -D_GNU_SOURCE -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS"
```
eventually we should get these from `llvm-config` from a HostDependency now that we have those

fingolfin replied:
Agreed! But i have no idea (yet) how to use those, and so I hope this can wait (or someone else does it)

fingolfin replied:
I've added a TODO comment stating this

vchuravy commented on L/libjulia/common.jl:
```diff
@@ -183,6 +198,8 @@ function build_julia(ARGS, version)
         JCPPFLAGS+=-DLIBOSXUNWIND
     EOM
+    else
+        echo "USECLANG=1" >>Make.user
```
Why is this needed? Can you add a comment.

fingolfin replied:
To be honest, I don't know -- I simply copied it from the script for Julia <= 1.6 above. I can try removing it to see what happens.

fingolfin replied:
Seems it works without this, so I'll remove it

vchuravy reviewed (commented):
Looks good from my side. 

The only question I had is why we don't list the version explicitly per dependency. E.g. if I try to rebuild libjulia 1.5, won't I get a version that is to new?

fingolfin commented:
For most dependencies, it doesn't matter which version we get, as they do not affect the Julia ABI, and that's all that counts for this JLL.

(That said, I wonder if not more of those deps should be turned into BuildDeps and/or be dropped because Julia now actually will fetch the relevant binaries anyway -- we only need to keep things that are needed to link against `libjulia` after all (i.e. dependencies of the shared library) and/or needed to include headers. But e.g. I don't see why `LibGit2_jll` would need to be a dependency.

Looking at `libjulia.dylib` and `libjulia-internal.dylib` on macOS, I see this:
```
$ otool -L usr/lib/libjulia.dylib
usr/lib/libjulia.dylib:
	@rpath/libjulia.dylib (compatibility version 1.0.0, current version 1.7.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1252.250.1)
$ otool -L usr/lib/libjulia-internal.dylib
usr/lib/libjulia-internal.dylib:
	@rpath/libjulia-internal.dylib (compatibility version 1.0.0, current version 1.7.0)
	@rpath/libjulia.dylib (compatibility version 1.0.0, current version 1.7.0)
	@rpath/libunwind.1.dylib (compatibility version 1.0.0, current version 1.0.0)
	@rpath/libLLVM.dylib (compatibility version 1.0.0, current version 12.0.0)
	/System/Library/Frameworks/CoreFoundation.framework/Versions/A/CoreFoundation (compatibility version 150.0.0, current version 1575.17.0)
	/usr/lib/libc++.1.dylib (compatibility version 1.0.0, current version 400.9.4)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1252.250.1)
```

Based on that, I wonder if we can't turn *all* dependencies into `BuildDependency` with the following exeptions:
- `libLLVM_jll` resp. `LLVM_full_jll`
- `LibUnwind_jll` resp. `LibOSXUnwind_jll` resp. `LLVMLibUnwind_jll`

But I am not 100% sure, perhaps I am missing something? I think to be sure, one also needs to experiment and see if a JLL build this way then can be used to e.g. build `libcxxwrap_julia`?

vchuravy commented:
I think that would actually be a step in the right direction. For libjulia we only want to define the dependencies the runtime itself uses and not the standard lib etc.

vchuravy commented on L/libjulia/common.jl:
```diff
@@ -15,6 +17,13 @@ function libjulia_platforms(julia_version)
+    # `utf8proc_jll` is missing the required artifacts for these architectures
+    if julia_version ≥ v"1.7"
+        filter!(p -> !(Sys.islinux(p) && arch(p) == "armv6l"), platforms)
+        filter!(p -> !(Sys.isapple(p) && arch(p) == "aarch64"), platforms)
```
```suggestion
```

fingolfin commented:
Thanks @giordano I've now enabled the "missing" ARM platforms and they all build! I've also changed most Deps into BuildDeps.
